SSA/Projects/RISCV64/Semantics.lean

```diff
@@ -89,83 +207,328 @@ def RTYPE_pure64_RISCV_XOR (rs2_val : BitVec 64) (rs1_val : BitVec 64) : BitVec
-def MULW_pure64 (rs2_val : (BitVec 64)) (rs1_val : BitVec 64) : BitVec 64 :=
+def REM_pure64_signed_bv (rs2_val : BitVec 64) (rs1_val : BitVec 64) : BitVec 64 :=
+  rs1_val.srem rs2_val
+
+theorem REM_pure64_signed_eq_REM_pure64_signed_bv (rs2_val : BitVec 64) (rs1_val : BitVec 64) :
+  REM_pure64_signed rs2_val rs1_val = REM_pure64_signed_bv rs2_val rs1_val := by
+  unfold REM_pure64_signed REM_pure64_signed_bv
+  rw [extractLsb'_ofInt_eq_ofInt (h:= by simp)]
+  split
+  case isTrue ht  =>
+    simp only [BitVec.ofInt_toInt, BitVec.srem, BitVec.umod_eq, BitVec.neg_eq]
+    obtain rfl : rs2_val = 0#_ :=
+      BitVec.eq_of_toInt_eq ht
+    split <;> simp
+  case isFalse hf =>
+    rw [← BitVec.toInt_srem, BitVec.ofInt_toInt]
+
+def MULW_pure64 (rs2_val : BitVec 64) (rs1_val : BitVec 64) : BitVec 64 :=
   BitVec.signExtend 64
     (BitVec.extractLsb 31 0
       (BitVec.extractLsb' 0 64
         (BitVec.ofInt 65 ((BitVec.extractLsb 31 0 rs1_val).toInt * (BitVec.extractLsb 31 0 rs2_val).toInt))))
+
+def MULW_pure64_bv (rs2_val : BitVec 64) (rs1_val : BitVec 64) : BitVec 64 :=
+  BitVec.signExtend 64
+    (((BitVec.extractLsb 31 0 rs1_val) * (BitVec.extractLsb 31 0 rs2_val)))
+
+theorem MULW_pure64_eq_MULW_pure64_bv (rs2_val : BitVec 64) (rs1_val : BitVec 64) :
+  MULW_pure64 rs2_val rs1_val = MULW_pure64_bv rs2_val rs1_val :=
+  by
+  unfold MULW_pure64 MULW_pure64_bv
+  rw [extractLsb'_ofInt_eq_ofInt (h:= by simp), BitVec.extractLsb]
+  simp only [Nat.sub_zero, Nat.reduceAdd]
+  rw [extractLsb'_ofInt_eq_ofInt (h:= by simp)]
+  congr
+  apply BitVec.eq_of_toInt_eq
+  simp only [BitVec.toInt_extractLsb, Nat.shiftRight_zero, Nat.sub_zero, Nat.reduceAdd,
+    Nat.reducePow, BitVec.toInt_ofInt, Int.mul_bmod_bmod, Int.bmod_mul_bmod, BitVec.toInt_mul]
+
 /-!
 ## mul operations flags
  the suffix indicates how the flags are assumed to be set.
 { high := _, signed_rs1:= _, signed_rs2 := _  }
 -/
 def MUL_pure64_fff (rs2_val : BitVec 64) (rs1_val : BitVec 64) : BitVec 64 :=
-  rs2_val * rs1_val
+  BitVec.extractLsb 63 0 (BitVec.extractLsb' 0 128 (BitVec.ofInt 129 (rs1_val.toNat * rs2_val.toNat)))
+
+def MUL_pure64_fff_bv (rs2_val : BitVec 64) (rs1_val : BitVec 64) : BitVec 64 :=  rs2_val * rs1_val
+
+theorem MUL_pure64_fff_eq_MUL_pure64_fff_bv (rs2_val : BitVec 64) (rs1_val : BitVec 64) :
+  MUL_pure64_fff  rs2_val rs1_val = MUL_pure64_fff_bv rs2_val rs1_val := by
+  simp only  [MUL_pure64_fff,  MUL_pure64_fff_bv]
+  apply BitVec.eq_of_toNat_eq
+  simp [Nat.sub_zero, Nat.reduceAdd, Int.mul_def, BitVec.extractLsb_toNat,
+    BitVec.extractLsb'_toNat, BitVec.toNat_ofInt, Nat.reducePow, Nat.cast_ofNat,
+    Nat.shiftRight_zero, Nat.reduceDvd, Nat.mod_mod_of_dvd, BitVec.mul_eq, BitVec.toNat_mul]
+  congr
+  norm_cast
+  rw [Int.toNat_natCast, Nat.mod_eq_of_lt]
+  ac_rfl
+  have aa := BitVec.isLt rs1_val
+  have bb := BitVec.isLt rs2_val
+  have := @Nat.mul_lt_mul'' _ _ _ _ aa bb
+  simp only [Nat.reducePow, Nat.reduceMul] at this
```

Comment:
Right before omega as the second-last tactic in a proof, you can (and should) also drop the `only`.